metric_learn/lmnn.py

```diff
@@ -101,233 +112,578 @@ class LMNN(MahalanobisMixin, TransformerMixin):
     super(LMNN, self).__init__(preprocessor)
 
   def fit(self, X, y):
-    k = self.k
-    reg = self.regularization
-    learn_rate = self.learn_rate
-
-    X, y = self._prepare_inputs(X, y, dtype=float,
-                                ensure_min_samples=2)
-    num_pts, d = X.shape
-    output_dim = _check_n_components(d, self.n_components)
-    unique_labels, label_inds = np.unique(y, return_inverse=True)
-    if len(label_inds) != num_pts:
-      raise ValueError('Must have one label per point.')
-    self.labels_ = np.arange(len(unique_labels))
-
-    self.components_ = _initialize_components(output_dim, X, y, self.init,
-                                              self.verbose,
-                                              random_state=self.random_state)
-    required_k = np.bincount(label_inds).min()
-    if self.k > required_k:
-      raise ValueError('not enough class labels for specified k'
-                       ' (smallest class has %d)' % required_k)
-
-    target_neighbors = self._select_targets(X, label_inds)
-
-    # sum outer products
-    dfG = _sum_outer_products(X, target_neighbors.flatten(),
-                              np.repeat(np.arange(X.shape[0]), k))
-
-    # initialize L
-    L = self.components_
-
-    # first iteration: we compute variables (including objective and gradient)
-    #  at initialization point
-    G, objective, total_active = self._loss_grad(X, L, dfG, k,
-                                                 reg, target_neighbors,
-                                                 label_inds)
-
-    it = 1  # we already made one iteration
+    """Train Large Margin Nearest Neighbor model.
+
+    :param X: array, shape (n_samples, n_features), the training samples.
+    :param y: array, shape (n_samples,), the training labels.
+
+    :return: The trained LMNN model.
+    """
+
+    # Check input arrays
+    X, y = self._prepare_inputs(X, y, dtype=float, ensure_min_samples=2)
+
+    n_samples, n_features = X.shape
+    n_components = _check_n_components(n_features, self.n_components)
+
+    # Initialize the random generator
+    self.random_state_ = check_random_state(self.random_state)
+
+    # Initialize transformation
+    self.components_ = _initialize_components(n_components, X, y, self.init,
+                                              verbose=self.verbose,
+                                              random_state=self.random_state_)
+
+    # remove singletons after initializing components
+    X, y, classes = self._validate_params(X, y)
+
+    # Find the target neighbors of each sample
+    target_neighbors = self._select_target_neighbors(X, y, classes)
+
+    # Compute the pull loss gradient w.r.t. M, which remains constant
+    tn_graph = _make_knn_graph(target_neighbors)
+    pull_loss_grad_m = _sum_weighted_outer_products(X, tn_graph)
+
+    # Compute the pull loss weight such that the push loss weight becomes 1
+    pull_loss_weight = (1 - self.push_loss_weight) / self.push_loss_weight
```

Comment:
The benefit is that you only need to multiply the pull loss by the rescaled weight once and the push loss never because its rescaled weight is one. I don't think the scaling has any negative effect in practice. If a user sets the `push_loss_weight=1e-5`, then she really intends for the pull loss to have such a higher weight relatively speaking.

Reply:
I will check but I think this is a bit complicated to change in the current code.

Reply:
I guess it is fine like this, but I think it could be good to at least advertise it in the docstring of the `push_loss_weight` parameter, that such a rescaling will happen ? 

Another possibility that may be good: I looked at our [User Guide](http://contrib.scikit-learn.org/metric-learn/supervised.html#lmnn) for LMNN, and in fact, as in the [Wikipedia page](https://en.wikipedia.org/wiki/Large_margin_nearest_neighbor#Algorithm) for LMNN, the weight on the **pull loss** is 1, and the weight on the **push loss** is set by the user (whereas in the [original paper](https://jmlr.csail.mit.edu/papers/volume10/weinberger09a/weinberger09a.pdf), they use weights that sum to 1)
So this means there's a bit of freedom for how people parameterize LMNN, so I'm thinking about this solution:
you could, instead of the `push_loss_term`, allow the user to set the `pull_loss_term`  weight(i.e. yet another formulation than Wikipedia or the original paper but they're all fine). Then in the docstring we would just say `pull_loss_term` is a coefficient in front of the pull loss term, (so not a relative weight), (which implicitly means the coefficient in front of the push loss term is 1). Doing that you would just have to replace the line of code above (i.e. `pull_loss_weight = (1 - self.push_loss_weight) / self.push_loss_weight`) by (`pull_loss_weight=self.pull_loss_weight`). The last thing would be just to update the User Guide by putting the coefficient `c` in front of the pull loss instead of the push loss.

I think that would preserve as much as possible your code while being very transparent and understandable (in the sense that the value of the cost function and gradient is exactly the one advertised in the User Guide, and there's no hidden rescaling)

What do you think ?

Reply:
I think this would be more confusing. The paper you cite is the journal paper, which was published a few years after the conference paper. The original conference paper (https://proceedings.neurips.cc/paper/2005/file/a7f592cef8b130a6967a90617db5681b-Paper.pdf) contains a formula for the cost function (Equation 2) with a push loss weight (`c`), and the formula in the User Guide is taken from there. So, I think keeping the `push_loss_weight` as a user parameter would be the most consistent option.